dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java

```diff
@@ -150,6 +150,8 @@ private LinkageChecker(
     
     for (ClassFile classFile : classToClassSymbols.keySet()) {
       ImmutableSet<ClassSymbol> classSymbols = classToClassSymbols.get(classFile);
```

Comment:
Not new in this PR, but I notice this is another case where the OO API is turned inside out. This should be `classFile.getClassSymbols`.

Reply:
The mismatch from OO API might came from my mental design of Linkage Checker:

- Get class path (JAR files) from input artifact
- Get symbol references in class files (such as this `classToClassSymbols`) from the class path
- Get linkage errors by validating the symbol references

I was feeling that this fits declarative/functional-programming style, not object-oriented:

```
linkageCheck = 
  artifact <- input
  classPath <- resolve(artifact)
  symbolReferences <- read(classPath)
  linkageErrors <- validate(symbolReferences)
  return linkageErrors
```

With that, I think I wanted to keep the clear separation of the value classes (e.g., `ClassFile`) from implementing IO operations. LinkageChecker and ClassDumper are responsible for the IO operations and logic to generate the value classes.

(This is not saying I'm against the object-oriented API, but sharing my finding.)

Reply:
Writing functional code in a non-functional language such as Java is a recipe for confusion. Case in point: I am constantly confused while reading our code. I usually have to review several classes to remind myself what the code does and how in each PR.

Separating I/O makes sense, but we shouldn't be storing the results of those I/O operations outside of the relevant classes.